pkg/controller/instance/instance_controller.go

```diff
@@ -264,11 +264,20 @@ func (r *Reconciler) Reconcile(request ctrl.Request) (ctrl.Result, error) {
 }
 
 func updateInstance(instance *kudov1beta1.Instance, oldInstance *kudov1beta1.Instance, client client.Client) error {
-	// update instance spec and metadata. this will not update Instance.Status field
+	// The order of both updates below is important: *first* the instance Spec and Metadata and *then* the Status.
+	// If Status is update first, a new reconcile request will be scheduled with *WRONG* instance snapshot (which is
+	// saved in the annotations). This request will then have wrong "previous state".
+
+	// 1. update instance spec and metadata. this will not update Instance.Status field
 	if !reflect.DeepEqual(instance.Spec, oldInstance.Spec) ||
-		!reflect.DeepEqual(instance.ObjectMeta.Annotations, oldInstance.ObjectMeta.Annotations) ||
-		!reflect.DeepEqual(instance.ObjectMeta.Finalizers, oldInstance.ObjectMeta.Finalizers) {
+		!reflect.DeepEqual(instance.ObjectMeta, oldInstance.ObjectMeta) {
+
 		instanceStatus := instance.Status.DeepCopy()
+
+		if tryRemoveFinalizer(instance) {
+			log.Printf("InstanceController: Removing finalizer on instance %s/%s", instance.Namespace, instance.Name)
```

Comment:
why is this not in the `tryRemoveFinalizer`?  seems if this is valuable and used other places that we would want to know for all cases...  what is logged is provided to the function.

Reply:
Good point, it's now there. That also allows us to distinguish the reasons why the finalizer was removed.

Reply:
Actually, #1352 moves this method (and it's adding-counterpart) into the helpers because they are not used by the webhook too. Now, `log.Printf("InstanceController:...` would be a lie. I think it is important to know who removed the finalizer so I kept it in the calling method. We could provide some logging prefix, of course, but that seems like overkill.

Reply:
On the other hand, I like the distinction between `cleanup plan is finished` and `there is no cleanup plan`. Ok, let's leave it like this.